Refactor TextIndicatorStyle to be seperate from UnifiedTextReplacementController.

megangardner opened the pull request:
#### 6b87830de4887034635dd02fdba85bdf0ebc18a6
<pre>
Refactor TextIndicatorStyle to be seperate from UnifiedTextReplacementController.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275169">https://bugs.webkit.org/show_bug.cgi?id=275169</a>
<a href="https://rdar.apple.com/129292403">rdar://129292403</a>

Reviewed by Aditya Keerthi.

TextIndicatorStyle is not always tied to what is happening in UnifiedTextReplacementController,
so it should be it&apos;s own separate system that interacts with UnifiedTextReplacementController
and WebPage to handle the current styling for all text indicators.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm: Added.
(WebKit::TextIndicatorStyleController::TextIndicatorStyleController):
(WebKit::TextIndicatorStyleController::document const):
(WebKit::TextIndicatorStyleController::contextRangeForTextIndicatorStyle const):
(WebKit::TextIndicatorStyleController::cleanUpTextStylesForSessionID):
(WebKit::TextIndicatorStyleController::removeTransparentMarkersForUUID):
(WebKit::newlyReplacedCharacterRange):
(WebKit::TextIndicatorStyleController::addSourceTextIndicatorStyle):
(WebKit::TextIndicatorStyleController::addDestinationTextIndicatorStyle):
(WebKit::TextIndicatorStyleController::updateTextIndicatorStyleVisibilityForID):
(WebKit::TextIndicatorStyleController::createTextIndicatorForRange):
(WebKit::TextIndicatorStyleController::createTextIndicatorForID):
(WebKit::TextIndicatorStyleController::enableTextIndicatorStyleAfterElementWithID):
(WebKit::TextIndicatorStyleController::enableTextIndicatorStyleForElementWithID):
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForUUID): Deleted.
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession): Deleted.
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
(WebKit::WebPage::enableTextIndicatorStyleAfterElementWithID):
(WebKit::WebPage::enableTextIndicatorStyleForElementWithID):
(WebKit::WebPage::getRangeForUUID): Deleted.
(WebKit::WebPage::createTextIndicatorForRange): Deleted.
* Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h: Added.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textIndicatorStyleController):
(WebKit::m_unifiedTextReplacementController): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::textIndicatorStyleController):
(WebKit::WebPage::unifiedTextReplacementController):

Canonical link: <a href="https://commits.webkit.org/279792@main">https://commits.webkit.org/279792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32bc45135c39b074a31a9d44bb8891c1407a50c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57845 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5298 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5300 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3568 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56659 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32135 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47259 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25317 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28930 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3438 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59435 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51618 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30947 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47345 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50995 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11926 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31925 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h:
```diff
@@ -0,0 +1,88 @@
+    RefPtr<WebCore::Document> document() const;
+    WeakPtr<WebPage> m_webPage;
+
+    HashMap<WTF::UUID, Vector<TextIndicatorStyleState>> m_activeTextIndicatorStyles;
```
this should be HashMap<UUID, HashMap<UUID, UniqueRef<CharacterRange>> (or unique_ptr)

rr-codes commented on Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h:
```diff
@@ -0,0 +1,88 @@
+
+    HashMap<WTF::UUID, Vector<TextIndicatorStyleState>> m_activeTextIndicatorStyles;
+    HashMap<WTF::UUID, WebCore::CharacterRange> m_currentlyStyledRange;
+    Vector<TextIndicatorStyleUnstyledRangeData> m_unstyledRanges;
```
can this be a hash map too?

rr-codes commented on Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h:
```diff
@@ -0,0 +1,88 @@
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+
+#include "UnifiedTextReplacementController.h"
```
is this import needed?

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
```
can this just be `RefPtr document = this->document`?

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+        return;
+    }
+
+    auto elementRange = makeRangeSelectingNodeContents(*element);
```
this should be `WebCore::` prefixed

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+    unstyledRange.start.container = replacedRangeAfterReplace.endContainer();
+    unstyledRange.start.offset = replacedRangeAfterReplace.endOffset();
+
+    auto unstyledRangeUUID = WTF::UUID::createVersion4();
```
can you use an ObjectIdentifier instead of a UUID?

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+        auto sessionRange = contextRangeForTextIndicatorStyle(uuid);
+
+        if (!sessionRange)
+            sessionRange = contextRangeForTextIndicatorStyle(uuid);
```
is this correct?

megangardner replied:
Nope, thanks!

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
```
can this use that `document()` method

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    WebCore::VisibleSelection fullDocumentSelection(WebCore::VisibleSelection::selectionFromContentsOfNode(root.get()));
+    auto simpleRange = fullDocumentSelection.range();
```
should this be `firstRange`

megangardner replied:
I don't know what you mean by this.

rr-codes replied:
instead of
```
auto simpleRange = fullDocumentSelection.range();
```


should it be
```
auto simpleRange = fullDocumentSelection.range();
```

? (which is what we do in UnifiedTextReplacementController

megangardner replied:
Those... are the same line... what did you mean for the second one?

rr-codes replied:
lol oops, I meant

```
auto simpleRange = fullDocumentSelection.firstRange();
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
```diff
@@ -481,89 +429,12 @@
-        m_textIndicatorCharacterRangesForSessions.append({ session.uuid, { { finalTextIndicatorUUID, characterRangeAfterReplace } } });
-#endif
-    }
+    m_webPage->textIndicatorStyleController().addDestinationTextIndicatorStyle(session.uuid, characterRangeAfterReplace, resolvedRange);
```
I think this should just call a method on WebPage that calls a method on `textIndicatorStyleController`, instead of exposing the controllers to each other

rr-codes commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -862,6 +863,14 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    UnifiedTextReplacementController& unifiedTextReplacementController() { return m_unifiedTextReplacementController.get(); };
+#endif
```
I think this should be avoided if possible

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,322 @@
+            if (styleState.styleID == uuid) {
+                auto fullSessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(sessionUUID);
+                if (fullSessionRange)
+                    return WebCore::resolveCharacterRange(*fullSessionRange, styleState.range, { WebCore::TextIteratorBehavior::EmitsObjectReplacementCharactersForImages });
```
this should use the same iterator behavior types that UTRC uses for consistency

megangardner replied:
They are the same, I just didn't make the same kind of helper you had. We need to make something for one or both of them to call.

rr-codes replied:
it's missing 

```
#if ENABLE(ATTACHMENT_ELEMENT)
    WebCore::TextIteratorBehavior::EmitsObjectReplacementCharactersForAttachments
#endif
```

megangardner replied:
Ahh, I think that happened while I was writing this patch, good catch :)

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,322 @@
+};
+#endif
+
+
```
extra newline

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,322 @@
+        return;
+    }
+
+    RefPtr root = document->documentElement();
```
this variable looks unused?

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,319 @@
+    } else {
+        Vector<TextIndicatorStyleState> styleStates = { styleState };
+        m_activeTextIndicatorStyles.add(sessionUUID, styleStates);
+    }
```
```suggestion
auto& styleStates = m_activeTextIndicatorStyles.ensure(sessionUUID, [&] {
  return Vector<TextIndicatorStyleState> { };
}).iterator->value;

styleStates.append(styleState);
```

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,319 @@
+        removeTransparentMarkersForUUID(styleState.styleID);
+
+    auto unstyledRange = m_unstyledRanges.find(sessionUUID);
+    removeTransparentMarkersForUUID(unstyledRange->value->styleID);
```
this should probably check that unstyledRange exists, and unstyledRange->value exists

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,319 @@
+
+std::optional<WebCore::SimpleRange> TextIndicatorStyleController::contextRangeForTextIndicatorStyle(const WTF::UUID& uuid) const
+{
+    auto sessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(uuid);
```
I don't think this variable is used anywhere other than the next two lines? Can move the assignment into the if.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,319 @@
+        for (auto styleState : m_activeTextIndicatorStyles.get(sessionUUID)) {
+            if (styleState.styleID == uuid) {
+                auto fullSessionRange = m_webPage->unifiedTextReplacementController().contextRangeForSessionWithUUID(sessionUUID);
+                if (fullSessionRange)
```
Same comment about moving the assignment into the if.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
```diff
@@ -0,0 +1,325 @@
+
+namespace WebKit {
+
+// This should mirror what in in UnifiedTextIndicatorController, and eventually not be copied.
```
```suggestion
// This should mirror what is in UnifiedTextIndicatorController, and eventually not be copied.
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h:
```diff
@@ -0,0 +1,91 @@
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+
+#include "WebPage.h"
```
I think this can be forward declared.

megangardner committed:
Refactor TextIndicatorStyle to be seperate from UnifiedTextReplacementController.
https://bugs.webkit.org/show_bug.cgi?id=275169
rdar://129292403

Reviewed by Aditya Keerthi.

TextIndicatorStyle is not always tied to what is happening in UnifiedTextReplacementController,
so it should be it's own separate system that interacts with UnifiedTextReplacementController
and WebPage to handle the current styling for all text indicators.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm: Added.
(WebKit::TextIndicatorStyleController::TextIndicatorStyleController):
(WebKit::TextIndicatorStyleController::document const):
(WebKit::TextIndicatorStyleController::contextRangeForTextIndicatorStyle const):
(WebKit::TextIndicatorStyleController::cleanUpTextStylesForSessionID):
(WebKit::TextIndicatorStyleController::removeTransparentMarkersForUUID):
(WebKit::newlyReplacedCharacterRange):
(WebKit::TextIndicatorStyleController::addSourceTextIndicatorStyle):
(WebKit::TextIndicatorStyleController::addDestinationTextIndicatorStyle):
(WebKit::TextIndicatorStyleController::updateTextIndicatorStyleVisibilityForID):
(WebKit::TextIndicatorStyleController::createTextIndicatorForRange):
(WebKit::TextIndicatorStyleController::createTextIndicatorForID):
(WebKit::TextIndicatorStyleController::enableTextIndicatorStyleAfterElementWithID):
(WebKit::TextIndicatorStyleController::enableTextIndicatorStyleForElementWithID):
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForUUID): Deleted.
(WebKit::UnifiedTextReplacementController::removeTransparentMarkersForSession): Deleted.
(WebKit::UnifiedTextReplacementController::contextRangeForSessionOrRangeWithUUID const): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createTextIndicatorForID):
(WebKit::WebPage::updateTextIndicatorStyleVisibilityForID):
(WebKit::WebPage::enableTextIndicatorStyleAfterElementWithID):
(WebKit::WebPage::enableTextIndicatorStyleForElementWithID):
(WebKit::WebPage::getRangeForUUID): Deleted.
(WebKit::WebPage::createTextIndicatorForRange): Deleted.
* Source/WebKit/WebProcess/WebPage/TextIndicatorStyleController.h: Added.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textIndicatorStyleController):
(WebKit::m_unifiedTextReplacementController): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::textIndicatorStyleController):
(WebKit::WebPage::unifiedTextReplacementController):

Canonical link: https://commits.webkit.org/279792@main
